core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,1099 @@
+    let failHandler = jasmine.createSpy('fail');
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId')
+      .and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+
+    expect(questionPlayerEngineService.getQuestionCount()).toBe(0);
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+
+    expect(questionPlayerEngineService.getQuestionCount()).toBe(3);
+  });
+
+  it('should update the exploration version if it ' +
+    'is not in question player mode when initialized', fakeAsync(() => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let exploration = sampleExplorationDict;
+    exploration.version = 2;
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(false);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+
+
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.callFake((expID, version) => {
+        expect(questionPlayerEngineService.getExplorationVersion()).toBe(1);
+        return Promise.resolve(exploration);
+      });
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+    tick();
+
+    expect(questionPlayerEngineService.getExplorationVersion()).toBe(2);
+  }));
+
+  it('should update the current question ID when a new card is added', () => {
+    let successCallback = jasmine.createSpy('success');
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let exploration = sampleExplorationDict;
+    let answer = 'answer';
+    let interactionRulesService: InteractionRulesService;
+    let answerClassificationResult = new AnswerClassificationResult(
+      outcomeObjectFactory
+        .createNew('default', '', '', []), 1, 0, 'default_outcome'
+    );
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.returnValue(Promise.resolve(exploration));
+    let answerClassificationServiceSpy =
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => {
```

Comment:
Probably can replace `{ return html; }`  with just `html`? Ditto elsewhere.

Reply:
Didn't get this part.
Are you suggesting to use `.and.returnValue(` ?

Reply:
See the second box below "arrow function without parameters" [here](https://www.tutorialsteacher.com/typescript/arrow-function) -- I just mean that I think you can simplify to `(html, envs) => html`.

Reply:
done.